util/gpool/spmc/spmspool.go

```diff
@@ -0,0 +1,386 @@
+	}
+}
+
+// Running returns the number of workers currently running.
+func (p *Pool[T, U, C, CT, TF]) Running() int {
+	return int(p.running.Load())
+}
+
+// Free returns the number of available goroutines to work, -1 indicates this pool is unlimited.
+func (p *Pool[T, U, C, CT, TF]) Free() int {
+	c := p.Cap()
+	if c < 0 {
+		return -1
+	}
+	return c - p.Running()
+}
+
+// Waiting returns the number of tasks which are waiting be executed.
+func (p *Pool[T, U, C, CT, TF]) Waiting() int {
+	return int(p.waiting.Load())
+}
+
+// IsClosed indicates whether the pool is closed.
+func (p *Pool[T, U, C, CT, TF]) IsClosed() bool {
+	return p.state.Load() == gpool.CLOSED
+}
+
+// Cap returns the capacity of this pool.
+func (p *Pool[T, U, C, CT, TF]) Cap() int {
+	return int(p.capacity.Load())
+}
+
+func (p *Pool[T, U, C, CT, TF]) addRunning(delta int) {
+	p.running.Add(int32(delta))
+}
+
+func (p *Pool[T, U, C, CT, TF]) addWaiting(delta int) {
+	p.waiting.Add(int32(delta))
+}
+
+// Release closes this pool and releases the worker queue.
+func (p *Pool[T, U, C, CT, TF]) Release() {
+	if !p.state.CompareAndSwap(gpool.OPENED, gpool.CLOSED) {
+		return
+	}
+	p.lock.Lock()
+	p.workers.reset()
+	p.lock.Unlock()
+	// There might be some callers waiting in retrieveWorker(), so we need to wake them up to prevent
+	// those callers blocking infinitely.
+	p.cond.Broadcast()
+}
+
+func isClose(exitCh chan struct{}) bool {
+	select {
+	case <-exitCh:
+		return true
+	default:
+	}
+	return false
+}
+
+// ReleaseAndWait is like Release, it waits all workers to exit.
+func (p *Pool[T, U, C, CT, TF]) ReleaseAndWait() {
+	if p.IsClosed() || isClose(p.stopCh) {
+		return
+	}
+
+	close(p.stopCh)
+	p.Release()
+	for {
+		if p.Running() == 0 && p.heartbeatDone.Load() {
+			return
+		}
+	}
+}
+
+// SetConsumerFunc is to set ConsumerFunc which is to process the task.
+func (p *Pool[T, U, C, CT, TF]) SetConsumerFunc(consumerFunc func(T, C, CT) U) {
+	p.consumerFunc = consumerFunc
+}
+
+// AddProduceBySlice is to add Produce by a slice.
+func (p *Pool[T, U, C, CT, TF]) AddProduceBySlice(producer func() ([]T, error), constArg C, contextFn TF, options ...TaskOption) (<-chan U, pooltask.TaskController[T, U, C, CT, TF]) {
+	opt := loadTaskOptions(options...)
+	taskID := p.NewTaskID()
+	var wg sync.WaitGroup
+	result := make(chan U, opt.ResultChanLen)
+	closeCh := make(chan struct{})
+	tc := pooltask.NewTaskController[T, U, C, CT, TF](p, taskID, closeCh, &wg)
+	taskCh := make(chan pooltask.Task[T], opt.TaskChanLen)
+	for i := 0; i < opt.Concurrency; i++ {
+		err := p.run()
+		if err == gpool.ErrPoolClosed {
+			break
+		}
+		taskBox := pooltask.NewTaskBox[T, U, C, CT, TF](constArg, contextFn, &wg, taskCh, result, taskID)
+		p.taskCh <- &taskBox
+	}
+	go func() {
+		defer func() {
+			close(closeCh)
+			close(taskCh)
+		}()
+		for {
+			tasks, err := producer()
+			if err != nil {
+				return
+			}
+			for _, task := range tasks {
+				wg.Add(1)
+				if err != nil {
```

Comment:
It doesn't feel like that logic(`err! = nil`) is happening here. Could we remove it?

Reply:
Done